lib/improver/cli/estimate_emos_coefficients.py

```diff
@@ -282,84 +99,66 @@ def process(historic_forecast, truth, combined, historic_forecast_dict,
-            msg = ("All of the combined_filepath, "
-                   "historic_forecast_identifier and truth_identifier "
-                   "arguments should be specified if one of the arguments are "
-                   "specified.")
-            raise ValueError(msg)
 
-    try:
-        if combined is not None:
-            historic_forecast, truth = SplitHistoricForecastAndTruth(
-                historic_forecast_dict, truth_dict).process(combined)
-    except ValueError as err:
-        # This error arises if the metadata to identify the desired historic
-        # forecast or truth has found nothing matching the metadata
-        # information supplied.
-        if str(err).startswith("The metadata to identify the desired"):
-            warnings.warn(str(err))
-            result = None
-        else:
-            raise
+    from collections import OrderedDict
+    from improver.utilities.cube_manipulation import MergeCubes
+    from improver.ensemble_calibration.ensemble_calibration import (
+        EstimateCoefficientsForEnsembleCalibration)
+
+    grouped_cubes = {}
+    for cube in cubes:
+        grouped_cubes.setdefault(cube.name(), []).append(cube)
+    if len(grouped_cubes) == 1:
+        # Only one group - all forecast/truth cubes
+        landsea_mask = None
+        diag_name = list(grouped_cubes.keys())[0]
+    elif len(grouped_cubes) == 2:
+        # Two groups - the one with exactly one cube matching a name should
+        # be the landmask, since we require >= 2 cubes in the forecast/truth
+        # group
+        grouped_cubes = OrderedDict(sorted(grouped_cubes.items(),
+                                           key=lambda kv: len(kv[1])))
+        # landsea name should be the key with the lowest number of cubes (1)
+        landsea_name, diag_name = list(grouped_cubes.keys())
+        landsea_mask = grouped_cubes[landsea_name][0]
+        if len(grouped_cubes[landsea_name]) != 1:
+            raise RuntimeError('Expected one cube for land-sea mask.')
     else:
-        result = EstimateCoefficientsForEnsembleCalibration(
-            distribution, cycletime, desired_units=units,
-            predictor_of_mean_flag=predictor_of_mean,
-            tolerance=tolerance, max_iterations=max_iterations).process(
-                historic_forecast, truth, landsea_mask=landsea_mask)
+        raise RuntimeError('Must have cubes with 1 or 2 distinct names.')
+
+    # split non-landmask cubes on forecast vs truth
+    truth_key, truth_value = truth_attribute.split('=')
+    input_cubes = grouped_cubes[diag_name]
+    grouped_cubes = {'truth': [], 'historical forecast': []}
+    for cube in input_cubes:
+        if cube.attributes.get(truth_key) == truth_value:
+            grouped_cubes['truth'].append(cube)
+        else:
+            grouped_cubes['historical forecast'].append(cube)
 
-    return result
+    missing_inputs = ' and '.join(k for k, v in grouped_cubes.items() if not v)
```

Comment:
Yes, with the current logic it is not possible to have `truth` and `historical forecast` both missing. However, if the logic for land-sea mask detection was changed (e.g. based on integer vs. float data type), then in principle this could happen. The logic for missing input detection is concise and self contained within its local block code, so I would leave it as is, unless you have a suggestion how to improve it.